docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+- Ubuntu does not provide the kernel module either but it can be installed using `apt-get install wireguard`. With that it appears straightforward to provide an image with WireGuard pre-installed.
+
+On the seed, we add a WireGuard device to one node on the host network. For all other nodes on the seed, we adapt the routes accordingly to route traffic destined for the WireGuard network to our WireGuard node. The Kubernetes pods managing the WireGuard device and routes are only used for initial configuration and later reconfiguration. During runtime, they can restart without any impact on the operation of the WireGuard network as the WireGuard device is managed by the Linux kernel.
+
+The WireGuard endpoint on the seed is exposed via a load balancer. We propose to use kubelink to manage the WireGuard configuration/device on the seed. We consider the management of the WireGuard endpoint to be complex especially in error situations which is the reason for utilizing kubelink as there is already significant experience managing an endpoint.
+
+On the shoot, we create the keys and acquire the WireGuard IP in the standard secret generation. The data is added as a secret to the control plane and to the shootstate. The vpn shoot deployment is extended to include the WireGuard device setup inside the vpn shoot pod network. For certain infrastructures (AWS), we need a re-advertiser to resolve the seed WireGuard endpoint and evaluate whether the IP address changed.
+
+The load balancer exposing the open vpn endpoint is no longer required and therefore removed if this functionality is used.
+
+As we want to slowly expand the usage of the WireGuard solution, we propose to introduce a feature gate for it. Furthermore, since the WireGuard network requires a separate network range, we propose to introduce a new section to the seed settings with two additional flags (enabled & cidr):
+
+```
+apiVersion: core.gardener.cloud/v1beta1
+kind: Seed
+metadata:
+  name: my-seed
+  ...
+spec:
+  ...
+  settings:
+  ...
+    wireguard:
+      enabled: true
+      cidr: 192.168.0.0/22
+```
+
+Last but not least, we propose to introduce an annotation to the shoots to enable/disable the WireGuard tunnel explicitly.
+
+```
+apiVersion: core.gardener.cloud/v1beta1
+kind: Shoot
+metadata:
+  name: my-shoot
+  annotations:
+    alpha.featuregates.shoot.gardener.cloud/wireguard-tunnel: "true"
+  ...
+```
+
+Using this approach, it is easy to switch the solution on and off, i.e. migrate the shoot clusters automatically during ordinary reconciliation.
+
+## High Availability
+
+There is an issue if the node that hosts the WireGuard endpoint fails. The endpoint is migrated to another node however the time required to do this might exceed the budget for downtimes although one could argue that a disruption of less than 30 seconds to 1 minute does not qualify as a downtime and will in almost all cases not noticeable by end users. 
```

Comment:
> 30 seconds to 1 minute does not qualify as a downtime 

It would be fine if this was true, but I was involved in discussions of less than 2 seconds downtime of CoreDNS of a shoot cluster. 1 minute downtime for the VPN potentially can block the kube apiservers due to admission webhooks relaying on connectivity from kube-apiserver to the in-cluster network. Multiplying by 200-300 shoots per seed, this is rather a disaster.

Reply:
We are following a different approach now.